src/twin/substrate.rs

```diff
@@ -89,11 +73,105 @@ where
     async fn get_twin_with_account(&self, account_id: AccountId32) -> Result<Option<u32>> {
         let mut client = self.client.lock().await;
 
+        let id = client.get_twin_id_by_account(account_id).await?;
+
+        Ok(id)
+    }
+}
+
+/// ClientWrapper is basically a substrate client.
+/// all methods exported by the ClientWrapper has a reconnect functionality internally.
+/// so if any network error happened, the ClientWrapper will try to reconnect to substrate using the provided substrate urls.
+/// if after a number of trials (currently 2 * the number of urls) a reconnect was not successful, the ClientWrapper returns an error.
+struct ClientWrapper {
+    client: Client,
+    substrate_urls: LinkedList<String>,
+}
+
+impl ClientWrapper {
+    pub async fn new(substrate_urls: Vec<String>) -> Result<Self> {
+        let mut urls = LinkedList::new();
+        for url in substrate_urls {
+            urls.push_back(url);
+        }
+
+        let client = Self::connect(&mut urls).await?;
+
+        Ok(Self {
+            client,
+            substrate_urls: urls,
+        })
+    }
+
+    async fn connect(urls: &mut LinkedList<String>) -> Result<Client> {
+        let trials = urls.len() * 2;
+        for _ in 0..trials {
+            let url = match urls.front() {
+                Some(url) => url,
+                None => anyhow::bail!("substrate urls list is empty"),
+            };
+
+            match Client::new(&url).await {
+                Ok(client) => return Ok(client),
+                Err(err) => {
+                    log::error!(
+                        "failed to create substrate client with url \"{}\": {}",
+                        url,
+                        err
+                    );
+                }
+            }
+
+            if let Some(front) = urls.pop_front() {
+                urls.push_back(front);
+            }
+        }
+
+        anyhow::bail!("failed to connect to substrate using the provided urls")
+    }
+
+    pub async fn update_twin(
+        &mut self,
+        kp: &KeyPair,
+        relay: Option<String>,
+        pk: Option<&[u8]>,
+    ) -> Result<()> {
+        let hash = loop {
+            match self.client.update_twin(kp, relay.clone(), pk).await {
+                Ok(hash) => break hash,
+                Err(ClientError::Rpc(_)) => {
+                    self.client = Self::connect(&mut self.substrate_urls).await?;
+                }
+                Err(err) => return Err(err.into()),
+            }
+        };
+
+        log::debug!("hash: {:?}", hash);
+
+        Ok(())
```

Comment:
this should return the hash, not log it. Then a caller to the update twin can choose either to update the hash or not!